File: Source/app/src/main/java/nl/endran/locust/game/GameLoop.java
/*
 * Copyright (c) 2015 by David Hardy. Licensed under the Apache License, Version 2.0.
 */

package nl.endran.locust.game;

import android.os.Handler;
import android.support.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

public class GameLoop {

    public interface Listener {
        void onTrigger(final int millisSincePreviousTrigger);
    }

    public static final int DELAY_MILLIS = 10;
    public static final int TRIGGER_INTERVAL_MILLIS = 100;

    @NonNull
    private final Handler handler;

    private final List<Listener> listeners = new ArrayList<>();
    private boolean isRunning;
    private long previousTimeMillis;

    @Inject
    public GameLoop(@NonNull final Handler handler) {
        this.handler = handler;
    }

    public void register(@NonNull final Listener listener) {
        listeners.add(listener);
        startLazy();
    }

    public void unregister(@NonNull final Listener listener) {
        listeners.remove(listener);
        stopLazy();
    }

    private void stopLazy() {
        if (listeners.isEmpty()) {
            isRunning = false;
        }
    }

    private void startLazy() {
        if (!isRunning) {
            isRunning = true;

            previousTimeMillis = System.currentTimeMillis();
            rePost();
        }
    }

    private void rePost() {
        handler.postDelayed(new Runnable() {
            @Override
            public void run() {
                if (!isRunning) {
                    return;
                }

                long currentTimeMillis = System.currentTimeMillis();
                long millisSincePreviousTrigger = currentTimeMillis - previousTimeMillis;
                if (millisSincePreviousTrigger > TRIGGER_INTERVAL_MILLIS) {
                    for (Listener listener : listeners) {
                        previousTimeMillis = currentTimeMillis;
                        listener.onTrigger((int) millisSincePreviousTrigger);
                    }
                }

                rePost();
            }
        }, DELAY_MILLIS);
    }
}

File: Source/app/src/main/java/nl/endran/locust/game/GameLoop.kt
/*
 * Copyright (c) 2015 by David Hardy. Licensed under the Apache License, Version 2.0.
 */

package nl.endran.locust.game

import android.os.Handler
import java.util.*
import javax.inject.Inject

class GameLoop
@Inject
constructor(private val handler: Handler) {

    interface Listener {
        fun onTrigger(millisSincePreviousTrigger: Int)
    }


    val DELAY_MILLIS = 10
    val TRIGGER_INTERVAL_MILLIS = 100

    private val listeners = ArrayList<Listener>()
    private var isRunning = false
    private var previousTimeMillis = 0L

    fun register(listener: Listener) {
        listeners.add(listener)
        startLazy()
    }

    fun unregister(listener: Listener) {
        listeners.remove(listener)
        stopLazy()
    }

    private fun stopLazy() {
        if (listeners.isEmpty()) {
            isRunning = false
        }
    }

    private fun startLazy() {
        if (!isRunning) {
            isRunning = true

            previousTimeMillis = System.currentTimeMillis()
            rePost()
        }
    }

    private fun rePost() {
        handler.postDelayed(Runnable {
            if (!isRunning) {
                return@Runnable
            }

            val currentTimeMillis = System.currentTimeMillis()
            val millisSincePreviousTrigger = currentTimeMillis - previousTimeMillis
            if (millisSincePreviousTrigger > TRIGGER_INTERVAL_MILLIS) {
                previousTimeMillis = currentTimeMillis
                val millisSincePreviousTriggerInt = millisSincePreviousTrigger.toInt()
                listeners.forEach { it.onTrigger(millisSincePreviousTriggerInt) }
            }

            rePost()
        }, DELAY_MILLIS.toLong())
    }
}
